File: 200-number-of-islands.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
  const knownSoil = new Set()
  const water = []
  let islands = 0

  if (!grid.length) {
    return 0
  }

  for (let y = 0; y < grid.length; y++) {
    for (let x = 0; x < grid[0].length; x++) {
      if (!isSoil(y, x)) {
        water.push([y, x])
      }
    }
  }

  if (!water.length) {
    return 1
  } else if (water.length === grid.length * grid[0].length) {
    return 0
  }

  while (water.length) {
    const [y, x] = water.pop()
    for (const [nextY, nextX] of nextMoves(y, x)) {
      if (isSoil(nextY, nextX)) {
        islands += exploreSoil(nextY, nextX)
      }
    }
  }

  function exploreSoil(y, x) {
    if (knownSoil.has(point(y, x))) {
      return 0
    }

    const soil = [[y, x]]

    while (soil.length) {
      const [y, x] = soil.pop()
      knownSoil.add(point(y, x))

      for (const [nextY, nextX] of nextMoves(y, x)) {
        if (isSoil(nextY, nextX)) {
          soil.push([nextY, nextX])
        }
      }
    }

    return 1
  }

  function point(y, x) {
    return Symbol.for(`${y},${x}`)
  }

  function isSoil(y, x) {
    // Implicit string to integer conversion
    return grid[y][x] == 1
  }

  function nextMoves(currentY, currentX) {
    const moves = [
      [-1, 0], [1, 0],
      [0, -1], [0, 1]
    ]

    return moves
      .map(([y, x]) => [y + currentY, x + currentX])
      .filter(isInGrid)
      .filter(([y, x])=> !knownSoil.has(point(y, x)))
  }

  function isInGrid([y, x]) {
    return y >= 0 && y < grid.length && x >= 0 && x < grid[0].length
  }

  return islands
}


const gridOne = [
  [1,1,0,0,0],
  [1,1,0,0,0],
  [0,0,1,0,0],
  [0,0,0,1,1]
]

const gridTwo = [
  [1,1,1,1,0],
  [1,1,0,1,0],
  [1,1,0,0,0],
  [0,0,0,0,0]
]

const gridThree = [
  [0,1,0,1,0],
  [0,0,1,0,0],
  [0,1,0,1,0],
  [0,0,1,0,0]
]

const gridFour = [
  [0,0,0,0,0],
  [0,1,1,1,0],
  [0,0,0,0,0],
  [0,0,1,0,0]
]

const gridFive = [
  [1,1,1,1,1],
  [1,1,1,1,1],
  [1,1,1,1,1],
  [1,1,1,1,1]
]

const gridSix = [
  [0,0,0,0,0],
  [0,0,0,0,0],
  [0,0,0,0,0],
  [0,0,0,0,0]
]

const failed = [
  ["1","1","1"],
  ["1","0","1"],
  ["1","1","1"]
]

console.log(numIslands(gridOne))
console.log(numIslands(gridTwo))
console.log(numIslands(gridThree))
console.log(numIslands(gridFour))
console.log(numIslands(gridFive))
console.log(numIslands(gridSix))
console.log(numIslands(failed))

File: 200-number-of-islands.py
from typing import List

'''
Given a 2d grid map of '1's (land) and '0's (water),
count the number of islands. An island is surrounded
by water and is formed by connecting adjacent lands
horizontally or vertically. You may assume all four
edges of the grid are all surrounded by water.
'''

class Solution:
  moves = {(-1,0), (0,1), (1,0), (0,-1)}

  def numIslands(self, grid: List[List[str]]) -> int:
    self.grid = grid
    self.width = range(len(grid))
    self.height = range(len(grid[0]))
    return self.find_islands()

  def get_grid(self, point: (int, int)):
    return self.grid[point[0]][point[1]]

  def in_grid(self, diff: (int, int)):
    return (diff[0] in self.width and diff[1] in self.height)

  def possible_moves(self, move: (int, int)):
    return {(move[0] + dy, move[1] + dx) for dy, dx in self.moves}

  def next(self, move: (int, int)):
    return set(filter(self.in_grid, self.possible_moves(move)))

  def find_islands(self):
    islands = 0

    known_soil = set()
    known_water = set()

    queue = [(0,0)]
    while len(queue):
      move = queue.pop(0)

      if self.get_grid(move) == 1:
        self.next(move)
        islands += 1
      else:
        pass

    return islands

# Solution ends
def print_grid(grid):
  for row in grid:
    print(row)
  print('<< end grid')

example_grid = [
  [1,1,0,0,0,0],
  [1,1,0,0,0,0],
  [0,0,1,0,0,0],
  [0,0,0,1,1,1]
]

sol = Solution()
result = sol.numIslands(example_grid)

print(result)
print(sol.known_soil)

# This is to ensure we check the grid correctly.
assert sol.in_grid((0,0)) == True
assert sol.in_grid((-1,0)) == False
assert sol.in_grid((0,-1)) == False
assert sol.in_grid((-1, -1)) == False
assert sol.in_grid((3,5)) == True
assert sol.in_grid((3,6)) == False
assert sol.in_grid((4,5)) == False
assert sol.in_grid((4,6)) == False